src/databricks/labs/ucx/source_code/graph.py

```diff
@@ -16,10 +17,12 @@ def __init__(
         dependency: Dependency,
         parent: DependencyGraph | None,
         resolver: DependencyResolver,
+        syspath_provider: SysPathProvider,
     ):
         self._dependency = dependency
         self._parent = parent
         self._resolver = resolver
+        self._syspath_provider = syspath_provider
```

Comment:
```suggestion
        self._syspath_provider = syspath_provider
        syspath_provider.push_cwd(dependency.path.parent)
```

this resolves hidden downstream bugs and prevents the need to:

1. change the interface from source container.
2. push/pop cwd in graph building implementations, that we may miss. 

<img width="612" alt="image" src="https://github.com/databrickslabs/ucx/assets/259697/385fa312-ce5d-4809-8d98-954eb35de2fb">

P.S. shouldn't `SysPathProvider` be renamed to `RelativePaths`, `PathLookup` or something like it?...

Reply:
I like `PathLookup`

Reply:
Re `push_cwd`, I'm not sure if the traversal strictly follows the instantiation, so we'd have to test if it can work. And as discussed on slack I don't think it can work with `sys.path`. (I appreciate this PR does not restrict itself to changes required by `cwd`, I was heading for `sys.path` and discovered the need to handle `cwd` first).
I'd rather implement `sys.path` fully (next PR) and then simplify if that's ok with you.

Reply:
it's incorrect to rely on `os.getcwd()` here, as it'll lead to undefined behavior.

https://github.com/databrickslabs/ucx/pull/1608#pullrequestreview-2036583040